### PR TITLE
Support GEMM + Swiglu fused MLP

### PR DESCRIPTION
**Pull request by ksivaman**
# What does this PR do ?

This PR supports GEMM + Swiglu fused MLP via Transformer Engine sequential ops.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/transformer/moe/experts.py [hunk 1]
```diff
@@ -1,12 +1,14 @@
 # Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 from __future__ import annotations
 
+from collections.abc import Callable
```
Comment (QiZhangNV) [inline megatron/core/transformer/moe/experts.py]: Duplicate import. Please remove one.

File: megatron/core/transformer/moe/moe_utils.py [hunk 2]
```diff
@@ -1286,9 +1286,11 @@ def get_align_size_for_quantization(config: TransformerConfig) -> int:
         int: The alignment size for quantization.
     """
     if config.fp8:
-        return get_fp8_align_size(config.fp8_recipe)
+        # CUTLASS kernel for MXFP8 grouped GEMM assumes 256 alignment
+        return 256
```
Comment (QiZhangNV) [inline megatron/core/transformer/moe/moe_utils.py]: Hardcoding alignment to 256 here affects all callers , not just the fused grouped GEMM path.  Please keep the original recipe-based alignment as default and only apply 256 when config.use_transformer_engine_op_fuser is enabled.

File: megatron/core/transformer/moe/experts.py [hunk 3]
```diff
@@ -737,6 +756,191 @@ def _apply_bias(intermediate_parallel, bias_parallel, tokens_per_expert, permute
...
+        # Apply padding if needed
+        unpadded_tokens_per_expert = None
+        padding_align_size = None
+        if self.config.moe_router_padding_for_quantization:
+            # Padding has already been applied in router
+            pass
+        elif self.config.fp8 or self.config.fp4:
+            tokens_per_expert = tokens_per_expert.tolist()
+            unpadded_tokens_per_expert = tokens_per_expert
+            padding_align_size = 256  # Requirement for grouped GEMM
```
Comment (QiZhangNV) [inline megatron/core/transformer/moe/experts.py]: Since Fp8Padding.__init__ already accepts align_size, it's better to set it at construction (line 732) instead:

```
align_size = 256 if self._with_fused_impl else None
self.quantization_padding = Fp8Padding(self.num_local_experts, align_size=align_size)
self.quantization_unpadding = Fp8Unpadding(self.num_local_experts, align_size=align_size)
```

File: megatron/core/transformer/moe/experts.py [hunk 4]
```diff
@@ -737,6 +756,191 @@ def _apply_bias(intermediate_parallel, bias_parallel, tokens_per_expert, permute
...
+
+
+        # Copy the weights from GroupedLinear module to GroupedLinear op.
+        if fc2_single_grouped_parameter:
+            setattr(op, "weight", getattr(self.linear_fc2, "weight"))
+
+        for idx in range(self.linear_fc2.num_gemms):
+            if not fc2_single_grouped_parameter:
+                setattr(op, f"weight{idx}", getattr(self.linear_fc2, f"weight{idx}"))
+            if self.linear_fc1.use_bias:
```
Comment (QiZhangNV) [inline megatron/core/transformer/moe/experts.py]:
```suggestion
            if self.linear_fc2.use_bias:
```

File: megatron/core/transformer/moe/experts.py [hunk 5]
```diff
@@ -737,6 +756,191 @@ def _apply_bias(intermediate_parallel, bias_parallel, tokens_per_expert, permute
             .to(intermediate_parallel.dtype)
         )
 
+    def _is_fused_impl_supported(self) -> bool:
```
Comment (QiZhangNV) [inline megatron/core/transformer/moe/experts.py:763]: `use_transformer_engine_op_fuser`  is defined but never checked. The fused path activates automatically whenever conditions are met. Please add:

```
if not self.config.use_transformer_engine_op_fuser:
    return False
```

File: megatron/core/transformer/transformer_config.py [hunk 6]
```diff
@@ -761,6 +765,15 @@ class TransformerConfig(ModelParallelConfig):
     """Number of SMs to use for HybridEP. In pure NVL scenarios,
     16 SMs can generally achieve good bandwidth."""
 
+    moe_mlp_glu_interleave_size: Optional[int] = None
```
Comment (QiZhangNV) [inline megatron/core/transformer/transformer_config.py:768]: This value is not validated. Please add an assert to reject invalid values in `TEGroupedMLP`.

File: megatron/core/transformer/moe/experts.py [hunk 7]
```diff
@@ -790,15 +1009,38 @@ def forward(
...
 
         def bias_act_func(intermediate_parallel, bias_parallel, permuted_probs):
+
+            # Whether activation function is interleaved GLU
+            with_glu_interleaving = (
+                self.config.gated_linear_unit
+                and self.config.moe_mlp_glu_interleave_size is not None
+            )
+
+            def remove_glu_interleaving(x: torch.Tensor) -> torch.Tensor:
```
Comment (QiZhangNV) [inline megatron/core/transformer/moe/experts.py:1022]: `remove_glu_interleaving`  runs a `contiguous()` copy every forward pass in the non-fused path. If the interleaved format is designed for the fused kernel, using it without the fused path is likely overhead with no benefit. Could we consider warning when moe_mlp_glu_interleave_size is set but _is_fused_impl_supported() returns False?

File: megatron/core/transformer/moe/experts.py [hunk 8]
```diff
@@ -737,6 +756,191 @@ def _apply_bias(intermediate_parallel, bias_parallel, tokens_per_expert, permute
...
+        # Note: The fused impl is stored in a tuple to avoid
+        # registering submodules.
+        if self._fused_ops is None:
+            self._fused_ops = (self._make_fused_ops(),)
+        ops, = self._fused_ops
+
+        # Apply padding if needed
+        unpadded_tokens_per_expert = None
+        padding_align_size = None
+        if self.config.moe_router_padding_for_quantization:
```
Comment (QiZhangNV) [inline megatron/core/transformer/moe/experts.py:918]: A quick question: for sync-free / full-iteration CUDA graph scenarios, is `moe_router_padding_for_quantization=True `always expected? In other words, the elif fallback exists because the fused ops are device-initiated, so padding must go through `D2H -> pad -> H2D `when not pre-padded in the router, making it not sync-free / CUDA-graphable?

File: megatron/core/transformer/transformer_config.py [hunk 9]
```diff
@@ -395,6 +395,10 @@ class TransformerConfig(ModelParallelConfig):
     fused_single_qkv_rope: bool = False
     """If set, avoid splitting QKV before ROPE forward and avoid concatenating ROPE dgrads."""
 
+    use_transformer_engine_op_fuser: bool = False
```
[the remark this answers is not in the document]
Comment (ksivaman) [reply megatron/core/transformer/transformer_config.py:398]: Done

File: megatron/core/transformer/moe/experts.py [hunk 10]
```diff
@@ -737,6 +756,191 @@ def _apply_bias(intermediate_parallel, bias_parallel, tokens_per_expert, permute
...
+            permuted_local_hidden_states,
+            tokens_per_expert,  # FC1
+            permuted_probs,  # Scaled SwiGLU
+            tokens_per_expert,  # FC2
+        )
+
+        # Remove padding if needed
+        if unpadded_tokens_per_expert is not None:
+            self.quantization_unpadding.align_size = padding_align_size
+            output = self.quantization_unpadding(output, unpadded_tokens_per_expert)
```
[the remark this answers is not in the document]
Comment (ksivaman) [reply megatron/core/transformer/moe/experts.py:945]: Addressed in above comment from @QiZhangNV

File: megatron/core/transformer/moe/experts.py [hunk 11]
```diff
@@ -1,12 +1,14 @@
 # Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 from __future__ import annotations
 
+from collections.abc import Callable
 import copy
 import logging
 from collections.abc import Callable
```
[the remark this answers is not in the document]
Comment (ksivaman) [reply megatron/core/transformer/moe/experts.py]: Addressed

File: megatron/core/transformer/moe/experts.py (hunk 5 above)
Comment (ksivaman) [reply megatron/core/transformer/moe/experts.py:763]: Good point, I have kept the function the same but instead added this check in the init as that made more sense

Comment (QiZhangNV) [reply megatron/core/transformer/moe/experts.py:763]: Thanks, looks good! A suggestion: since the assert above already guards _is_fused_impl_supported(), maybe line 753 could be simplified to self._with_fused_impl = self.config.use_transformer_engine_op_fuser?

File: megatron/core/transformer/moe/experts.py (hunk 8 above)
Comment (ksivaman) [reply megatron/core/transformer/moe/experts.py:918]: Yes, that's the expect path

File: megatron/core/transformer/moe/experts.py [hunk 12]
```diff
@@ -712,10 +747,24 @@ def __init__(
 
             set_save_original_input(self.linear_fc1)
 
+        # Fused implementation with Transformer Engine op fuser API
+        if self.config.use_transformer_engine_op_fuser:
+            assert self._is_fused_impl_supported(), "Fused GroupedMLP is not supported for this configuration."
+        self._with_fused_impl: bool = self._is_fused_impl_supported() and self.config.use_transformer_engine_op_fuser
```
Comment (ksivaman) [inline megatron/core/transformer/moe/experts.py]:
```suggestion
        self._with_fused_impl: bool = self.config.use_transformer_engine_op_fuser
```

File: megatron/core/transformer/moe/experts.py (hunk 5 above)
Comment (ksivaman) [reply megatron/core/transformer/moe/experts.py:763]: Done

**Comment by yaox12** [comment]
Can you add unit tests for the numerics of the fusion, and the remapping of parameter keys?

**Comment by ksivaman** [comment]
@yaox12 The unit test for both the checkpoint loading as well as specifically the fusion numerics are included in https://github.com/NVIDIA/TransformerEngine/pull/2769. I think adding them separately would be a duplicate. Unless you mean an e2e test?

File: megatron/core/transformer/moe/experts.py [hunk 13]
```diff
@@ -133,6 +152,19 @@ def activation_func_with_probs(x, probs):
         # How many feature each rank holds for fc1 and fc2, respectively.
         tp_size = self.tp_group.size()
         tp_rank = self.tp_group.rank()
+        if (
```
Comment (zhongbozhu) [inline megatron/core/transformer/moe/experts.py]: Why do we need to edit class GroupedMLP when we are actually using TEGroupedMLP?

File: megatron/core/transformer/moe/experts.py [hunk 14]
```diff
@@ -712,10 +744,32 @@ def __init__(
...
+                "tensor reordering/copy overhead each forward pass.",
+                self.config.moe_mlp_glu_interleave_size,
+            )
+
         if self.config.fp8 or self.config.fp4:
             assert HAVE_TE, "FP8 and FP4 requires TE."
-            self.quantization_padding = Fp8Padding(self.num_local_experts)
-            self.quantization_unpadding = Fp8Unpadding(self.num_local_experts)
+            align_size = 256 if self._with_fused_impl else None
+            self.quantization_padding = Fp8Padding(self.num_local_experts, align_size=align_size)
```
Comment (zhongbozhu) [inline megatron/core/transformer/moe/experts.py:738]: So the Fp8Padding module itself is not graph safe, when we are running full CG, this module should be disabled and let permute kernel do the padding, will this be address in a followup PR?

File: megatron/core/transformer/moe/experts.py [hunk 15]
```diff
@@ -737,6 +791,192 @@ def _apply_bias(intermediate_parallel, bias_parallel, tokens_per_expert, permute
...
+        return True
+
+    def _make_fused_ops(self) -> torch.nn.Module:
+        """Construct fused module for FC1, activation, and FC2."""
+
+        # Container for fusible ops
+        ops = te.pytorch.ops.Sequential()
+
+        # Check if there are 1 or "num_gemms" params in the GroupedLinear module.
+        fc1_single_grouped_parameter = self.linear_fc1.single_grouped_parameter
```
Comment (zhongbozhu) [inline megatron/core/transformer/moe/experts.py:803]: Is single grouped weight still being controlled by a TE env var?

File: megatron/core/transformer/moe/experts.py (hunk 13 above)
Comment (yaox12) [reply megatron/core/transformer/moe/experts.py]: You're right. In fact the whole `GroupedMLP` class has been removed on the main branch, which has not been synced to the dev branch yet.

File: megatron/core/transformer/moe/experts.py [hunk 16]
```diff
@@ -737,6 +791,192 @@ def _apply_bias(intermediate_parallel, bias_parallel, tokens_per_expert, permute
...
+        # the params are modified after the constructor.
+        # Note: The fused impl is stored in a tuple to avoid
+        # registering submodules.
+        if self._fused_ops is None:
+            self._fused_ops = (self._make_fused_ops(),)
+        (ops,) = self._fused_ops
+
+        # Apply padding if needed
+        unpadded_tokens_per_expert = None
+        if self.config.moe_router_padding_for_quantization:
```
Comment (zhongbozhu) [inline megatron/core/transformer/moe/experts.py:918]: both router padding & explicit padding are not graph safe, only padding with permute kernels are graph safe, will this be addressed in the followup PR?

File: megatron/core/transformer/moe/experts.py [hunk 17]
```diff
@@ -737,6 +791,192 @@ def _apply_bias(intermediate_parallel, bias_parallel, tokens_per_expert, permute
...
+            # Padding has already been applied in router
+            pass
+        elif self.config.fp8 or self.config.fp4:
+            tokens_per_expert = tokens_per_expert.tolist()
```
Comment (zhongbozhu) [inline megatron/core/transformer/moe/experts.py:922]: Same^, this will be fixed in the next PR?
